modules/runtime.go

```diff
@@ -0,0 +1,283 @@
+	err := ctx.PushBusy()
+	defer ctx.PopBusy()
+	if err != nil {
+		return nil, err
+	}
+
+	if impl.Code == nil && len(impl.CodeSrc) > 0 {
+		impl.Code, err = py.Compile(string(impl.CodeSrc), impl.Info.FileDesc, py.ExecMode, 0, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if impl.Code == nil && len(impl.CodeBuf) > 0 {
+		codeBuf := bytes.NewBuffer(impl.CodeBuf)
+		obj, err := marshal.ReadObject(codeBuf)
+		if err != nil {
+			return nil, err
+		}
+		impl.Code, _ = obj.(*py.Code)
+		if impl.Code == nil {
+			return nil, py.ExceptionNewf(py.AssertionError, "Embedded code did not produce a py.Code object")
+		}
+	}
+
+	module, err := ctx.Store().NewModule(ctx, impl)
+	if err != nil {
+		return nil, err
+	}
+
+	if impl.Code != nil {
+		_, err = ctx.RunCode(impl.Code, module.Globals, module.Globals, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return module, nil
+}
+
+func (ctx *context) ResolveAndCompile(pathname string, opts py.CompileOpts) (py.CompileOut, error) {
+	err := ctx.PushBusy()
+	defer ctx.PopBusy()
+	if err != nil {
+		return py.CompileOut{}, err
+	}
+
+	tryPaths := defaultPaths
+	if opts.UseSysPaths {
+		tryPaths = ctx.Store().MustGetModule("sys").Globals["path"].(*py.List).Items
+	}
+
+	out := py.CompileOut{}
+
+	err = resolveRunPath(pathname, opts, tryPaths, func(fpath string) (bool, error) {
+
+		stat, err := os.Stat(fpath)
+		if err == nil && stat.IsDir() {
+			// FIXME this is a massive simplification!
+			fpath = path.Join(fpath, "__init__.py")
+			_, err = os.Stat(fpath)
+		}
+
+		ext := strings.ToLower(filepath.Ext(fpath))
+		if ext == "" && os.IsNotExist(err) {
+			fpath += ".py"
+			ext = ".py"
+			_, err = os.Stat(fpath)
+		}
+
+		// Keep searching while we get FNFs, stop on an error
+		if err != nil {
+			if os.IsNotExist(err) {
+				return true, nil
+			}
+			err = py.ExceptionNewf(py.OSError, "Error accessing %q: %v", fpath, err)
+			return false, err
+		}
+
+		switch ext {
+		case ".py":
+			var pySrc []byte
+			pySrc, err = ioutil.ReadFile(fpath)
+			if err != nil {
+				return false, py.ExceptionNewf(py.OSError, "Error reading %q: %v", fpath, err)
+			}
+
+			out.Code, err = py.Compile(string(pySrc), fpath, py.ExecMode, 0, true)
+			if err != nil {
+				return false, err
+			}
+			out.SrcPathname = fpath
+		case ".pyc":
+			file, err := os.Open(fpath)
+			if err != nil {
+				return false, py.ExceptionNewf(py.OSError, "Error opening %q: %v", fpath, err)
+			}
+			defer file.Close()
+			codeObj, err := marshal.ReadPyc(file)
+			if err != nil {
+				return false, py.ExceptionNewf(py.ImportError, "Failed to marshal %q: %v", fpath, err)
+			}
+			out.Code, _ = codeObj.(*py.Code)
+			out.PycPathname = fpath
+		}
+
+		out.FileDesc = fpath
+		return false, nil
+	})
+
+	if out.Code == nil && err == nil {
+		err = py.ExceptionNewf(py.AssertionError, "Missing code object")
+	}
+
+	if err != nil {
+		return py.CompileOut{}, err
+	}
+
+	return out, nil
+}
+
+func (ctx *context) PushBusy() error {
```

Comment:
it seems to `PushBusy` is only used inside the `modules` package.
so we can probably just not export it.
```suggestion
func (ctx *context) pushBusy() error {
```